File: csharp/BronKerbosch/Reporter.cs
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Diagnostics;
using Vertex = System.UInt32;

namespace BronKerbosch
{
    public interface Reporter
    {
        void Record(ImmutableArray<Vertex> clique);
    }

    public sealed class SimpleReporter : Reporter
    {
        public List<ImmutableArray<Vertex>> Cliques { get; } = new List<ImmutableArray<Vertex>>();

        public void Record(ImmutableArray<Vertex> clique)
        {
            Debug.Assert(clique.Length > 1);
            Cliques.Add(clique);
        }
    }

    public sealed class CountingReporter : Reporter
    {
        public int Cliques { get; private set; } = 0;

        public void Record(ImmutableArray<Vertex> clique)
        {
            Cliques += 1;
        }
    }
}

File: go/src/BronKerbosch/reporter.go
package BronKerbosch

type Reporter interface {
	Record(clique []Vertex)
}

type SimpleReporter struct {
	Cliques [][]Vertex
}

func (r *SimpleReporter) Record(clique []Vertex) {
	cc := make([]Vertex, len(clique))
	copy(cc, clique)
	r.Cliques = append(r.Cliques, cc)
}

type CountingReporter struct {
	Cliques int
}

func (r *CountingReporter) Record(clique []Vertex) {
	r.Cliques += 1
}

type ChannelReporter struct {
	cliques chan<- []Vertex
}

func (r *ChannelReporter) Record(clique []Vertex) {
	cc := make([]Vertex, len(clique))
	copy(cc, clique)
	r.cliques <- cc
}

func gatherCliques(cliques <-chan []Vertex, finalReporter Reporter) {
	for clique := range cliques {
		finalReporter.Record(clique)
	}
}
